File: python-solutions/rotate-image.py
from typing import List


class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        l, r = 0, len(matrix) - 1

        while l < r:
            for i in range(r - l):
                top, bottom = l, r

                # save the top left
                top_left = matrix[top][l + i]

                # move the bottom left into the top left
                matrix[top][l + i] = matrix[bottom - i][l]

                # move the bottom right into the bottom left
                matrix[bottom - i][l] = matrix[bottom][r - i]

                # move top right into the bottom right
                matrix[bottom][r - i] = matrix[top + i][r]

                # move the top left temp var into the top right position
                matrix[top + i][r] = top_left
            r -= 1
            l += 1

File: javascript-solutions/rotate-image.js
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var rotate = function (matrix) {
	let l = 0,
		r = matrix[0].length - 1;
	while (l < r) {
		for (let i = 0; i < r - l; i++) {
			const top = l,
				bottom = r;

			// store the top left value as a temp value
			const topLeft = matrix[top][l + i];

			// move bottom left to top left
			matrix[top][l + i] = matrix[bottom - i][l];

			// move bottom right to bottom left
			matrix[bottom - i][l] = matrix[bottom][r - i];

			// move top right to bottom right
			matrix[bottom][r - i] = matrix[top + i][r];

			// move the top left temp value to top right
			matrix[top + i][r] = topLeft;
		}
		r--;
		l++;
	}
};
